05-2 교차 검증과 그리드 서치

In [3]:
#데이터를 불러옵니다
import pandas as pd 

wine = pd.read_csv('https://bit.ly/wine_csv_data')

In [5]:
#class 열을 타깃으로 사용하고 나머지 열은 특성 배열에 저장 

data = wine[['alcohol','sugar','pH']].to_numpy()
target = wine['class'].to_numpy()

In [7]:
#훈련 세트의 입력 데이터와 타깃 데이터를 train_input, train_target 배열에 저장 
from sklearn.model_selection import train_test_split

train_input, test_input, train_target, test_target = train_test_split(
    data, target, test_size=0.2, random_state=42
)

In [8]:
#val 설정해주기 

sub_input, val_input, sub_target, val_target= train_test_split(
    train_input, train_target, test_size=0.2, random_state=42
)

In [9]:
# 훈련 세트와 검증 세트의 크기를 확인 

print(sub_input.shape, val_input.shape)

(4157, 3) (1040, 3)


원래 있던 훈련세트 5197개가 4157개로 줄고 
검증세트는 1040개가 되었음 

In [10]:
#모델을 만들고 평가 

from sklearn.tree import DecisionTreeClassifier
dt = DecisionTreeClassifier(random_state=42)
dt.fit(sub_input, sub_target)
print(dt.score(sub_input, sub_target))
print(dt.score(val_input, val_target))

0.9971133028626413
0.864423076923077


이 모델은 훈련세트에 과대적합되었기에 매개변수를 바꾸어서 더 좋은 모델을 찾아야한다! 